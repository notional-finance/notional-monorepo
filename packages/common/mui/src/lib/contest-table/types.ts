import { MessageDescriptor } from 'react-intl';
import { Column } from 'react-table';

export enum CONTEST_TABLE_VARIANTS {
  COMPACT = 'compact',
  DEFAULT = 'default',
}



export type ContestTableColumn = Column & {
  textAlign?: string;
  showLinkIcon?: boolean;
  defaultCanSort?: boolean;
  sortType?: string;
  sortDescFirst?: boolean;
  displayFormatter?: any; 
  showSymbol?: boolean; 
  columnHeaderToolTip?: MessageDescriptor; 
  marginRight?: any; 
  isIDCell?: boolean;
  padding?: string;
  display?: string;
};
