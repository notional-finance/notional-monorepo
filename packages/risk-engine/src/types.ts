import { TokenBalance, TokenDefinition } from '@notional-finance/core-entities';

/** Denomination can be either a specific currency id or symbol */
export type SymbolOrID = string | number;

export interface RiskFactors {
  // Absolute values
  netWorth: () => TokenBalance;
  freeCollateral: () => TokenBalance;
  // Ratios
  loanToValue: () => number;
  collateralRatio: () => number | null;
  healthFactor: () => number | null;
  leverageRatio: () => number | null;
  // Thresholds
  liquidationPrice: (
    debt: TokenDefinition,
    collateral: TokenDefinition
  ) => TokenBalance | null;
  collateralLiquidationThreshold: (
    collateral: TokenDefinition
  ) => TokenBalance | null;
}

export interface RiskFactorLimit<F extends keyof RiskFactors> {
  riskFactor: F;
  limit: NonNullable<ReturnType<RiskFactors[F]>>;
  args?: Parameters<RiskFactors[F]>;
}
