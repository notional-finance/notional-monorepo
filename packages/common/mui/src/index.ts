export * from './lib/action-sidebar/action-sidebar';
export * from './lib/action-sidebar-buttons/action-sidebar-buttons';
export * from './lib/area-chart/area-chart';
export * from './lib/bar-chart/bar-chart';
export * from './lib/asset-select-dropdown/asset-select-dropdown';
export * from './lib/app-bar/app-bar';
export * from './lib/box-display/box-display';
export * from './lib/button/button';
export * from './lib/button-bar/button-bar';
export * from './lib/card/card';
export * from './lib/chart-tool-tip/chart-tool-tip';
export * from './lib/chart-header/chart-header';
export * from './lib/chart-header-totals/chart-header-totals';
export * from './lib/chart-info-box/chart-info-box';
export * from './lib/copy-caption/copy-caption';
export * from './lib/count-up/count-up';
export * from './lib/countdown-cards/countdown-cards';
export * from './lib/currency-img/currency-img';
export * from './lib/currency-input/currency-input';
export * from './lib/data-table/types';
export * from './lib/data-table/data-table';
export * from './lib/infinite-scroll-data-table/infinite-scroll-data-table';
export * from './lib/data-table-cells';
export * from './lib/contest-table/contest-table';
export * from './lib/contest-table/contest-cells';
export * from './lib/contest-table/types';
export * from './lib/drawer/drawer';
export * from './lib/dropdown/dropdown';
export * from './lib/dropdown-button/dropdown-button';
export * from './lib/error-message/error-message';
export * from './lib/discord-button-gradient/discord-button-gradient';
export * from './lib/faq/faq';
export * from './lib/faq-header/faq-header';
export * from './lib/hooks';
export * from './lib/info-tooltip/info-tooltip';
export * from './lib/input/input';
export * from './lib/input-label/input-label';
export * from './lib/interactive-area-chart/interactive-area-chart';
export * from './lib/lever-up-toggle/lever-up-toggle';
export * from './lib/light-dark-switch/light-dark-switch';
export * from './lib/hooks/use-window-dimensions/use-window-dimensions';
export * from './lib/labeled-text/labeled-text';
export * from './lib/mini-button/mini-button';
export * from './lib/multi-display-chart/multi-display-chart';
export * from './lib/multi-select-dropdown/multi-select-dropdown';
export * from './lib/page-loading/page-loading';
export * from './lib/progress-indicator/progress-indicator';
export * from './lib/product-dashboard/product-dashboard';
export * from './lib/section/section';
export * from './lib/scroll-to-top/scroll-to-top';
export * from './lib/select-dropdown/select-dropdown';
export * from './lib/section-link/section-link';
export * from './lib/side-bar-layout/side-bar-layout';
export * from './lib/slider-input/slider-input';
export * from './lib/slider-display/slider-display';
export * from './lib/slider-basic/slider-basic';
export * from './lib/slider-risk/slider-risk';
export * from './lib/side-bar-sub-header/side-bar-sub-header';
export * from './lib/side-drawer/side-drawer';
export * from './lib/side-drawer-button/side-drawer-button';
export * from './lib/side-drawer-active-button/side-drawer-active-button';
export * from './lib/switch/switch';
export * from './lib/tab-toggle/tab-toggle';
export * from './lib/toggle/toggle';
export * from './lib/toggle-switch/toggle-switch';
export * from './lib/total-box/total-box';
export * from './lib/total-box/total-row';
export * from './lib/trade-action-header/trade-action-header';
export * from './lib/trade-action-title/trade-action-title';
export * from './lib/trade-summary-container/trade-summary-container';
export * from './lib/trade-summary-box/trade-summary-box';
export * from './lib/note-chart/note-chart';
export * from './lib/type-form/type-form';
export * from './lib/typography/typography';
export * from './lib/external-link/external-link';
export * from './lib/drawer-transition/drawer-transition';
export * from './lib/manage-side-drawer/manage-side-drawer';
export type { CurrencySelectOption } from './lib/currency-input/currency-select';
export * from './lib/simple-toggle/simple-toggle';
export * from './lib/footer-popup/footer-popup';
export * from './lib/message-box/message-box';
export * from './lib/network-toggle/network-toggle';
export * from './lib/mobile-nav-with-filter/mobile-nav-with-filter';
