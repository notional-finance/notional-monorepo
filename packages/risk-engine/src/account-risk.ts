import {
  Registry,
  TokenBalance,
  TokenDefinition,
} from '@notional-finance/core-entities';
import {
  Network,
  PERCENTAGE_BASIS,
  RATE_PRECISION,
} from '@notional-finance/util';
import { BaseRiskProfile } from './base-risk';
import { SymbolOrID } from './types';

export class AccountRiskProfile extends BaseRiskProfile {
  static from(balances: TokenBalance[]) {
    return new AccountRiskProfile(balances);
  }

  static simulate(from: TokenBalance[], apply: TokenBalance[]) {
    return new AccountRiskProfile([...from, ...apply]);
  }

  simulate(apply: TokenBalance[]) {
    return AccountRiskProfile.simulate(this.balances, apply);
  }

  /** Takes a set of token balances to create a new account risk profile */
  constructor(_balances: TokenBalance[], _network?: Network) {
    const balances = _balances.filter(
      (t) =>
        !t.isVaultToken &&
        t.tokenType !== 'Underlying' &&
        t.tokenType !== 'Fiat'
    );

    super(balances, 'ETH', _network);
  }

  protected _totalRiskAdjusted(b: TokenBalance[], d: TokenDefinition) {
    return this._totalValue(
      b.map((t) => t.toToken(d, t.isNegative() ? 'Debt' : 'Asset')),
      d
    );
  }

  /***** RISK RATIOS *******/
  collateralRatio() {
    const debts = this.totalDebt();
    if (debts.isZero()) return null;

    return this._toPercent(this.totalAssets(), debts.neg());
  }

  maxLoanToValue() {
    const ltv = this.loanToValue();
    const assets = this.totalAssetsRiskAdjusted();
    if (assets.isZero()) return 0;

    const riskAdjustedLTV = this._toPercent(
      this.totalDebtRiskAdjusted(),
      assets
    );

    return ltv / riskAdjustedLTV;
  }

  /** Total value of all assets with a risk adjustment */
  totalAssetsRiskAdjusted(denominated = this.defaultSymbol) {
    return this._totalRiskAdjusted(this.collateral, this.denom(denominated));
  }

  /** Total debt with risk adjustments */
  totalDebtRiskAdjusted(denominated = this.defaultSymbol) {
    return this._totalRiskAdjusted(this.debts, this.denom(denominated));
  }

  /** Total value of assets in the specified currency */
  totalCurrencyAssets(currencyId: number, denominated = this.defaultSymbol) {
    return this._totalValue(
      this.collateral.filter((t) => t.token.currencyId === currencyId),
      this.denom(denominated)
    );
  }

  /** Total value of assets in the specified currency with risk adjustments */
  totalCurrencyAssetsRiskAdjusted(currencyId: number) {
    return this._totalRiskAdjusted(
      this.collateral.filter((t) => t.token.currencyId === currencyId),
      Registry.getTokenRegistry().getPrimeCash(this.network, currencyId)
    );
  }

  /** Total value of debts in the specified currency with risk adjustments */
  totalCurrencyDebtsRiskAdjusted(currencyId: number) {
    return this._totalRiskAdjusted(
      this.debts.filter((t) => t.token.currencyId === currencyId),
      Registry.getTokenRegistry().getPrimeCash(this.network, currencyId)
    );
  }

  /**
   * Net value of debts and assets in the specified token with risk adjustments, denominated
   * in the given token.
   */
  netCollateralAvailable(_collateral: SymbolOrID) {
    const collateral = this.denom(_collateral);

    return this._totalRiskAdjusted(
      this.balances.filter((t) =>
        // If underlying is specified the sum up all assets in that underlying
        collateral.tokenType === 'Underlying'
          ? t.token.underlying === collateral.id
          : t.token.id === collateral.id
      ),
      collateral
    );
  }

  freeCollateralFactors() {
    const denom = this.denom(this.defaultSymbol);

    return this.allCurrencyIds.map((currencyId) => {
      const totalAssetsLocal = this.totalCurrencyAssetsRiskAdjusted(currencyId);
      const totalDebtsLocal = this.totalCurrencyDebtsRiskAdjusted(currencyId);
      const netLocal = totalAssetsLocal.add(totalDebtsLocal);
      let netFC: TokenBalance;
      let totalAssetsETH: TokenBalance;
      let totalDebtsETH: TokenBalance;

      if (netLocal.currencyId === denom.currencyId) {
        // Handle special case when netLocal is in the denom currency
        const { haircut, buffer } =
          Registry.getConfigurationRegistry().getCurrencyHaircutAndBuffer(
            denom
          );
        netFC = netLocal
          .toUnderlying()
          .scale(netLocal.isPositive() ? haircut : buffer, PERCENTAGE_BASIS);
        totalAssetsETH = totalAssetsLocal
          .toUnderlying()
          .scale(haircut, PERCENTAGE_BASIS);
        totalDebtsETH = totalDebtsLocal
          .toUnderlying()
          .scale(buffer, PERCENTAGE_BASIS);
      } else {
        netFC = netLocal.toToken(
          denom,
          netLocal.isNegative() ? 'Debt' : 'Asset'
        );
        totalAssetsETH = totalAssetsLocal.toToken(denom, 'Asset');
        totalDebtsETH = totalDebtsLocal.toToken(denom, 'Debt');
      }

      return {
        currencyId,
        netFC,
        totalAssetsETH,
        totalDebtsETH,
      };
    });
  }

  healthFactor() {
    // As defined here: https://docs.notional.finance/notional-v2/borrower-resources/health-factor
    const denom = this.denom(this.defaultSymbol);
    const factors = this.freeCollateralFactors();
    const totalAssets = factors.reduce(
      (t, { totalAssetsETH }) => t.add(totalAssetsETH),
      TokenBalance.zero(denom)
    );
    const totalDebts = factors.reduce(
      (t, { totalDebtsETH }) => t.add(totalDebtsETH),
      TokenBalance.zero(denom)
    );
    return totalDebts.isZero()
      ? null
      : totalAssets.toFloat() / totalDebts.abs().toFloat();
  }

  freeCollateral() {
    const denom = this.denom(this.defaultSymbol);
    const factors = this.freeCollateralFactors();
    return factors.reduce(
      (fc, { netFC }) => fc.add(netFC),
      TokenBalance.zero(denom)
    );
  }

  leverageRatio(currencyId?: number): number | null {
    let totalAssets: TokenBalance;
    let totalDebt: TokenBalance;
    if (currencyId) {
      // If currencyId is specified, then the leverage ratio
      // is specific to the denomination currency, without cross
      // currency haircuts applied.
      totalAssets = this.totalCurrencyAssetsRiskAdjusted(currencyId);
      totalDebt = this.totalCurrencyDebtsRiskAdjusted(currencyId);
    } else {
      totalDebt = this.totalDebt();
      totalAssets = this.totalAssets();
    }

    return totalDebt.isZero()
      ? null
      : totalDebt.neg().ratioWith(totalAssets.add(totalDebt)).toNumber() /
          RATE_PRECISION;
  }

  /***** RISK THRESHOLD *******/
  assetLiquidationThreshold(asset: TokenDefinition) {
    let riskAdjustedValue = this.netCollateralAvailable(asset.id);
    // If there is no collateral available, then the liquidation price is null
    if (this.totalDebt().isZero()) return null;
    if (riskAdjustedValue.isZero()) return null;
    const { haircut, buffer } =
      Registry.getConfigurationRegistry().getCurrencyHaircutAndBuffer(asset);
    let assetDenominatedFC: TokenBalance;

    if (asset.tokenType === 'Underlying') {
      if (riskAdjustedValue.isPositive() && haircut === 0) {
        // No cross currency liquidation price if token is haircut to zero
        return null;
      } else {
        // Apply haircut on cross currency risk
        riskAdjustedValue = riskAdjustedValue.scale(
          riskAdjustedValue.isPositive() ? haircut : buffer,
          PERCENTAGE_BASIS
        );
        assetDenominatedFC = this.freeCollateral().toToken(asset);
      }
    } else {
      // In this case we're determining asset price risk and the "FC" figure is the
      // total asset's contribution to the free collateral (i.e. the net collateral
      // available of the asset's underlying.
      if (!asset.underlying) throw Error('No underlying found');
      assetDenominatedFC = this.netCollateralAvailable(asset.underlying);
      assetDenominatedFC = assetDenominatedFC.scale(
        assetDenominatedFC.isPositive() ? haircut : buffer,
        PERCENTAGE_BASIS
      );
    }

    const unitOfAsset = TokenBalance.unit(asset);

    // 1 - collateralDenominatedFC / netCollateralAvailable
    // If netCollateralAvailable < 0 then the exchange rate change will be positive
    const threshold = unitOfAsset.sub(
      unitOfAsset.scale(assetDenominatedFC, riskAdjustedValue)
    );

    if (asset.tokenType === 'nToken') {
      const { nTokenMaxDrawdown } =
        Registry.getConfigurationRegistry().getNTokenLeverageFactors(asset);

      // This is the minimum price for an nToken
      if (threshold.lt(unitOfAsset.mulInRatePrecision(nTokenMaxDrawdown)))
        return null;
    } else if (asset.tokenType === 'fCash') {
      if (riskAdjustedValue.isPositive()) {
        const { lowestDiscountFactor } =
          Registry.getConfigurationRegistry().getMinLendRiskAdjustedDiscountFactor(
            asset
          );

        // This is the minimum price for positive fCash
        if (threshold.lt(unitOfAsset.mulInRatePrecision(lowestDiscountFactor)))
          return null;
      } else {
        // This is the maximum price for fCash debt
        if (threshold.gt(unitOfAsset)) return null;
      }
    }

    if (
      (threshold.isNegative() || threshold.isZero()) &&
      assetDenominatedFC.isPositive()
    ) {
      // If the max exchange rate decrease is negative then there is no possible liquidation price, this can
      // happen if aggregateFC > netUnderlying.
      return null;
    }

    return threshold;
  }

  borrowCapacity(currencyId: number) {
    const usedBorrowCapacity = this.totalCurrencyDebts(currencyId);
    const netCollateralAvailable = this.netCollateralAvailable(
      usedBorrowCapacity.token.id
    );
    const fcInDebtDenomination = this.freeCollateral().toToken(
      usedBorrowCapacity.token
    );
    const debtBuffer = Registry.getConfigurationRegistry().getDebtBuffer(
      this.network,
      currencyId
    );

    let additionalBorrowCapacity: TokenBalance;
    if (netCollateralAvailable.isPositive()) {
      if (netCollateralAvailable.gt(fcInDebtDenomination)) {
        // Limiting factor is FC, so reduce it to zero
        additionalBorrowCapacity = fcInDebtDenomination;
      } else {
        // Reduces netCollateralAvailable to zero as well as whatever additional FC is available
        const remainingFC = fcInDebtDenomination
          .sub(netCollateralAvailable)
          .scale(100, debtBuffer);
        additionalBorrowCapacity = netCollateralAvailable.add(remainingFC);
      }
    } else {
      // Any additional borrow capacity will come with a buffer attached
      additionalBorrowCapacity = fcInDebtDenomination.scale(100, debtBuffer);
    }

    return {
      usedBorrowCapacity,
      additionalBorrowCapacity,
      totalBorrowCapacity: usedBorrowCapacity.add(additionalBorrowCapacity),
    };
  }

  getAllRiskFactors() {
    return {
      netWorth: this.netWorth(),
      freeCollateral: this.freeCollateral(),
      loanToValue: this.loanToValue(),
      collateralRatio: this.collateralRatio(),
      leverageRatio: this.leverageRatio(),
      healthFactor: this.healthFactor(),
      liquidationPrice: this.getAllLiquidationPrices(),
    };
  }

  override maxWithdraw(token: TokenDefinition): TokenBalance {
    const balance = this.collateral.find((t) => t.token.id === token.id);
    if (!balance) return TokenBalance.zero(token);

    // If the token's value is haircut to zero then we still have to take into
    // account net local
    const { haircut } =
      Registry.getConfigurationRegistry().getCurrencyHaircutAndBuffer(
        balance.underlying
      );

    if (haircut === 0) {
      const netLocal = this.totalCurrencyAssetsRiskAdjusted(balance.currencyId)
        .add(this.totalCurrencyDebtsRiskAdjusted(balance.currencyId))
        .toToken(token);

      return netLocal.lt(balance) ? netLocal : balance;
    }

    const maxWithdraw = this.getWithdrawRequiredToMaintainRiskFactor(token, {
      riskFactor: 'freeCollateral',
      limit: TokenBalance.zero(this.denom(this.defaultSymbol)),
    });

    return maxWithdraw.neg().lt(balance) ? maxWithdraw.neg() : balance;
  }
}
