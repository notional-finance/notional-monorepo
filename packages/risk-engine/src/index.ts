export * from './types';
export { AccountRiskProfile } from './account-risk';
import { BaseRiskProfile } from './base-risk';

export type RiskProfile = typeof BaseRiskProfile;
