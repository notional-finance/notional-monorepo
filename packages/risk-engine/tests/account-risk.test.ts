import {
  AccountFetchMode,
  Registry,
  TokenBalance,
} from '@notional-finance/core-entities';
import { Network } from '@notional-finance/util';
import { AccountRiskProfile } from '../src/account-risk';
import { RiskFactorKeys, RiskFactorLimit, RiskFactors } from '../src/types';

interface Profile {
  name: string;
  balances: [number, string][];
  expected: {
    factor: keyof RiskFactors;
    args?: string[];
    expected: [number, string] | number | null;
  }[];
}

describe.withForkAndRegistry(
  {
    network: Network.ArbitrumOne,
    fetchMode: AccountFetchMode.SINGLE_ACCOUNT_DIRECT,
  },
  'Account Risk',
  () => {
    const profiles: Profile[] = [
      {
        name: 'One ETH',
        balances: [[1, 'ETH']],
        expected: [
          { factor: 'netWorth', expected: [1, 'ETH'] },
          { factor: 'freeCollateral', expected: [0.81, 'ETH'] },
          { factor: 'loanToValue', expected: 0 },
          { factor: 'collateralRatio', expected: null },
          { factor: 'healthFactor', expected: null },
          {
            factor: 'collateralLiquidationThreshold',
            args: ['ETH'],
            expected: null,
          },
        ],
      },
      {
        name: 'ETH / USDC',
        balances: [
          [1, 'ETH'],
          [-1280, 'USDC'],
        ],
        expected: [
          { factor: 'netWorth', expected: [0.324, 'ETH'] },
          { factor: 'freeCollateral', expected: [0.263, 'ETH'] },
          { factor: 'loanToValue', expected: 67.59 },
          { factor: 'collateralRatio', expected: 147.94 },
          { factor: 'healthFactor', expected: 0.8122 },
          {
            factor: 'collateralLiquidationThreshold',
            args: ['ETH'],
            expected: [0.736, 'ETH'],
          },
          {
            factor: 'liquidationPrice',
            args: ['USDC', 'ETH'],
            expected: [1395.2, 'USDC'],
          },
        ],
      },
      {
        name: 'Net Prime Cash',
        balances: [
          [-5, 'pdEther'],
          [10, 'pEther'],
        ],
        expected: [
          { factor: 'netWorth', expected: [5, 'ETH'] },
          { factor: 'freeCollateral', expected: [4.05, 'ETH'] },
          { factor: 'loanToValue', expected: 0 },
          { factor: 'collateralRatio', expected: null },
          { factor: 'healthFactor', expected: null },
        ],
      },
      {
        name: 'Leveraged nToken',
        balances: [
          [-8, 'pdEther'],
          [10, 'nETH'],
        ],
        expected: [
          { factor: 'netWorth', expected: [1.996, 'ETH'] },
          { factor: 'freeCollateral', expected: [0.996, 'ETH'] },
          { factor: 'loanToValue', expected: 80.03 },
          { factor: 'collateralRatio', expected: 124.94 },
          { factor: 'healthFactor', expected: 0.499 },
          { factor: 'leverageRatio', expected: 4.008 },
        ],
      },
    ];

    const riskFactors: {
      riskFactor: RiskFactorKeys;
      limit: number | [number, string];
      args?: [string];
    }[] = [
      {
        riskFactor: 'loanToValue',
        limit: 60,
      },
      {
        riskFactor: 'loanToValue',
        limit: 70,
      },
      {
        riskFactor: 'collateralRatio',
        limit: 140,
      },
      {
        riskFactor: 'collateralRatio',
        limit: 160,
      },
      {
        riskFactor: 'netWorth',
        limit: [0.31, 'ETH'],
      },
      {
        riskFactor: 'netWorth',
        limit: [0.37, 'ETH'],
      },
      {
        riskFactor: 'freeCollateral',
        limit: [0.2, 'ETH'],
      },
      {
        riskFactor: 'freeCollateral',
        limit: [0.35, 'ETH'],
      },
      {
        riskFactor: 'collateralLiquidationThreshold',
        args: ['ETH'],
        limit: [0.65, 'ETH'],
      },
      {
        riskFactor: 'collateralLiquidationThreshold',
        args: ['ETH'],
        limit: [0.85, 'ETH'],
      },
    ];

    it.each(profiles)('Risk Factors: $name', ({ name, balances, expected }) => {
      const tokens = Registry.getTokenRegistry();
      const p = AccountRiskProfile.from(
        balances.map(([n, t]) =>
          TokenBalance.fromFloat(
            n,
            tokens.getTokenBySymbol(Network.ArbitrumOne, t)
          )
        )
      );

      expected.forEach(({ factor, expected, args = [] }) => {
        const _args = args.map((t: string) =>
          tokens.getTokenBySymbol(Network.ArbitrumOne, t)
        );
        const actual = p.getRiskFactor(factor as keyof RiskFactors, _args);
        const logResults = name.includes('LOG');
        if (Array.isArray(expected)) {
          const e = TokenBalance.fromFloat(
            expected[0],
            tokens.getTokenBySymbol(Network.ArbitrumOne, expected[1])
          );

          if (logResults) {
            console.log(
              `${name} / ${factor}: ${(
                actual as TokenBalance
              ).toDisplayStringWithSymbol(8)} ${e.toDisplayStringWithSymbol(8)}`
            );
          } else {
            expect(actual).toBeApprox(e, 1e-2);
          }
        } else if (expected === null) {
          if (logResults) {
            console.log(
              `${name} / ${factor}: ${
                actual instanceof TokenBalance
                  ? actual.toDisplayStringWithSymbol(8)
                  : actual
              } null`
            );
          } else {
            expect(actual).toBe(null);
          }
        } else {
          if (logResults) {
            console.log(`${name} / ${factor}: ${actual} ${expected}`);
          } else {
            expect(actual).toBeCloseTo(expected, 1);
          }
        }
      });
    });

    it.each(riskFactors)(
      'Deposit / Withdraw Maintain Factor [$riskFactor | $limit]',
      ({ riskFactor, limit, args }) => {
        const tokens = Registry.getTokenRegistry();
        const ETH = tokens.getTokenBySymbol(Network.ArbitrumOne, 'ETH');
        const USDC = tokens.getTokenBySymbol(Network.ArbitrumOne, 'USDC');
        const p = AccountRiskProfile.from([
          TokenBalance.fromFloat(1, ETH),
          TokenBalance.fromFloat(-1280, USDC),
        ]);

        const l =
          typeof limit === 'number'
            ? limit
            : TokenBalance.fromFloat(
                limit[0],
                tokens.getTokenBySymbol(Network.ArbitrumOne, limit[1])
              );

        const _args = (
          args
            ? args.map((s) => tokens.getTokenBySymbol(Network.ArbitrumOne, s))
            : undefined
        ) as RiskFactorLimit<RiskFactorKeys>['args'];
        const deposit = p.getDepositRequiredToMaintainRiskFactor(ETH, {
          riskFactor,
          limit: l,
          args: _args,
        });

        if (typeof limit === 'number') {
          expect(
            p.simulate([deposit]).getRiskFactor(riskFactor, _args)
          ).toBeCloseTo(l, 1);
        } else {
          expect(
            p.simulate([deposit]).getRiskFactor(riskFactor, _args)
          ).toBeApprox(l, 0.01);
        }
      }
    );

    it.each(riskFactors.filter(({ riskFactor }) => riskFactor !== 'netWorth'))(
      'Leverage / Deleverage Maintain Factor [$riskFactor | $limit]',
      ({ riskFactor, limit, args }) => {
        const tokens = Registry.getTokenRegistry();
        const ETH = tokens.getTokenBySymbol(Network.ArbitrumOne, 'ETH');
        const USDC = tokens.getTokenBySymbol(Network.ArbitrumOne, 'USDC');
        const p = AccountRiskProfile.from([
          TokenBalance.fromFloat(1, ETH),
          TokenBalance.fromFloat(-1280, USDC),
        ]);
        const l =
          typeof limit === 'number'
            ? limit
            : TokenBalance.fromFloat(
                limit[0],
                tokens.getTokenBySymbol(Network.ArbitrumOne, limit[1])
              );

        const _args = (
          args
            ? args.map((s) => tokens.getTokenBySymbol(Network.ArbitrumOne, s))
            : undefined
        ) as RiskFactorLimit<RiskFactorKeys>['args'];
        const { netCollateral, netDebt } =
          p.getDebtAndCollateralMaintainRiskFactor(USDC, ETH, {
            riskFactor,
            limit: l,
            args: _args,
          });

        if (typeof limit === 'number') {
          expect(
            p
              .simulate([netCollateral, netDebt])
              .getRiskFactor(riskFactor, _args)
          ).toBeCloseTo(l, 0);
        } else {
          expect(
            p
              .simulate([netCollateral, netDebt])
              .getRiskFactor(riskFactor, _args)
          ).toBeApprox(l, 0.01);
        }
      }
    );

    it('All Liquidation Prices', () => {
      const tokens = Registry.getTokenRegistry();
      const ETH = tokens.getTokenBySymbol(Network.ArbitrumOne, 'ETH');
      const BTC = tokens.getTokenBySymbol(Network.ArbitrumOne, 'WBTC');
      const USDC = tokens.getTokenBySymbol(Network.ArbitrumOne, 'USDC');
      const FRAX = tokens.getTokenBySymbol(Network.ArbitrumOne, 'FRAX');
      const p = AccountRiskProfile.from([
        TokenBalance.fromFloat(1, ETH),
        TokenBalance.fromFloat(0.01, BTC),
        TokenBalance.fromFloat(-1200, USDC),
        TokenBalance.fromFloat(-100, FRAX),
      ]);

      const prices = p.getAllLiquidationPrices({ onlyUnderlyingDebt: false });
      const pairs = prices.map(({ collateral, debt }) => [
        collateral.symbol,
        debt.symbol,
      ]);

      expect(pairs).toEqual([
        ['ETH', 'USDC'],
        ['ETH', 'FRAX'],
      ]);
    });

    it('All Risk Factors', () => {
      const tokens = Registry.getTokenRegistry();
      const ETH = tokens.getTokenBySymbol(Network.ArbitrumOne, 'ETH');
      const BTC = tokens.getTokenBySymbol(Network.ArbitrumOne, 'WBTC');
      const USDC = tokens.getTokenBySymbol(Network.ArbitrumOne, 'USDC');
      const FRAX = tokens.getTokenBySymbol(Network.ArbitrumOne, 'FRAX');
      const p = AccountRiskProfile.from([
        TokenBalance.fromFloat(1, ETH),
        TokenBalance.fromFloat(0.01, BTC),
        TokenBalance.fromFloat(-1200, USDC),
        TokenBalance.fromFloat(-100, FRAX),
      ]);
      expect(p.getAllRiskFactors()).toBeDefined();
    });
  }
);
