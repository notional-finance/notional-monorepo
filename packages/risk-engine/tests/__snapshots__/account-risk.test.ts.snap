// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With Registry Account Risk Risk Factors: ETH / USDC 1`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x0439a3c132c4e800",
  "network": "arbitrum",
  "tokenId": "0x0000000000000000000000000000000000000000",
}
`;

exports[`With Registry Account Risk Risk Factors: ETH / USDC 2`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0xb83a3699ef4300",
  "network": "arbitrum",
  "tokenId": "0x0000000000000000000000000000000000000000",
}
`;

exports[`With Registry Account Risk Risk Factors: ETH / USDC 3`] = `69.5545496`;

exports[`With Registry Account Risk Risk Factors: ETH / USDC 4`] = `143.7720471`;

exports[`With Registry Account Risk Risk Factors: ETH / USDC 5`] = `1.0683977734379138`;

exports[`With Registry Account Risk Risk Factors: ETH / USDC 6`] = `2.284563013`;

exports[`With Registry Account Risk Risk Factors: ETH / USDC 7`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x0d287c7cee8c3b58",
  "network": "arbitrum",
  "tokenId": "0x0000000000000000000000000000000000000000",
}
`;

exports[`With Registry Account Risk Risk Factors: ETH / USDC 8`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x106579",
  "network": "arbitrum",
  "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
}
`;

exports[`With Registry Account Risk Risk Factors: ETH / USDC 9`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x68005a77",
  "network": "arbitrum",
  "tokenId": "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
}
`;

exports[`With Registry Account Risk Risk Factors: ETH / USDC 10`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x61af5c",
  "network": "arbitrum",
  "tokenId": "0xabc07bf91469c5450d6941dd0770e6e6761b90d6",
}
`;

exports[`With Registry Account Risk Risk Factors: Leveraged fCash Spread 1`] = `
Object {
  "_isTokenBalance": true,
  "hex": "-0x0135f25ae49c00",
  "network": "arbitrum",
  "tokenId": "0x0000000000000000000000000000000000000000",
}
`;

exports[`With Registry Account Risk Risk Factors: Leveraged fCash Spread 2`] = `0.813708434288799`;

exports[`With Registry Account Risk Risk Factors: Leveraged fCash Spread 3`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x0632af48",
  "network": "arbitrum",
  "tokenId": "0x0000000000000000000000000000000000000000000000000003006580dd0001",
}
`;

exports[`With Registry Account Risk Risk Factors: Leveraged nToken 1`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x1ba108c2185e9800",
  "network": "arbitrum",
  "tokenId": "0x0000000000000000000000000000000000000000",
}
`;

exports[`With Registry Account Risk Risk Factors: Leveraged nToken 2`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x0b23e5b94582f100",
  "network": "arbitrum",
  "tokenId": "0x0000000000000000000000000000000000000000",
}
`;

exports[`With Registry Account Risk Risk Factors: Leveraged nToken 3`] = `80.0881494`;

exports[`With Registry Account Risk Risk Factors: Leveraged nToken 4`] = `124.8624181`;

exports[`With Registry Account Risk Risk Factors: Leveraged nToken 5`] = `0.7340701832704267`;

exports[`With Registry Account Risk Risk Factors: Leveraged nToken 6`] = `4.022134911`;

exports[`With Registry Account Risk Risk Factors: Leveraged nToken 7`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x02a3040219a7b171",
  "network": "arbitrum",
  "tokenId": "0x0000000000000000000000000000000000000000",
}
`;

exports[`With Registry Account Risk Risk Factors: Leveraged nToken 8`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x069079d6",
  "network": "arbitrum",
  "tokenId": "0x18b0fc5a233acf1586da7c199ca9e3f486305a29",
}
`;

exports[`With Registry Account Risk Risk Factors: Leveraged nToken w/ Liquidation Risk 1`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x05b93d08",
  "network": "arbitrum",
  "tokenId": "0x18b0fc5a233acf1586da7c199ca9e3f486305a29",
}
`;

exports[`With Registry Account Risk Risk Factors: Leveraged nToken w/ Liquidation Risk 2`] = `
Object {
  "_isTokenBalance": true,
  "hex": "0x033fd4d5",
  "network": "arbitrum",
  "tokenId": "0x18b0fc5a233acf1586da7c199ca9e3f486305a29",
}
`;

exports[`With Registry Account Risk Risk Factors: Zero Haircut Collateral 1`] = `
Object {
  "_isTokenBalance": true,
  "hex": "-0xd542ebab6a7c00",
  "network": "arbitrum",
  "tokenId": "0x0000000000000000000000000000000000000000",
}
`;
